keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2508 @@
+fails, the scheduling attempt needs to be aborted and retried at a later time
+or when the ResourceClaims change.
+
+#### Unreserve
+
+The scheduler removes the Pod from the ReservedFor field because it cannot be scheduled after
+all.
+
+### Cluster Autoscaler
+
+When [Cluster
+Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#cluster-autoscaler)
+encounters a pod that uses a resource claim, the autoscaler needs assistance by
+the resource driver for that claim to make the right decisions. Without that
+assistance, the autoscaler might scale up the wrong node group (resource is
+provided by nodes in another group) or scale up unnecessarily (resource is
+network-attached and adding nodes won't help).
+
+At the start of a scale up or scale down cycle, autoscaler takes a snapshot of
+the current cluster state. Then autoscaler determines whether a real or
+fictional node fits a pod by calling the pre-filter and filter extension points
+of scheduler plugins. If a pod fits a node, the snapshot is updated by calling
+[NodeInfo.AddPod](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/framework/types.go#L620-L623). This
+influences further checks for other pending pods.
+
+To support the custom allocation logic that a vendor uses for its resources,
+the autoscaler needs an extension mechanism similar to the [scheduler
+extender](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/framework/extender.go#L24-L72). The
+existing scheduler extender API has to be extended to include methods that
+would only get called by the autoscaler, like starting a cycle. Instead of
+adding these methods to the scheduler framework, autoscaler can define its own
+interface that inherits from the framework:
+
+```
+import "k8s.io/pkg/scheduler/framework"
+
+type Extender interface {
+    framework.Extender
+
+    // NodeSelected gets called when the autoscaler determined that
+    // a pod should run on a node.
+    NodeSelected(pod *v1.Pod, node *v1.Node) error
+
+    // NodeReady gets called by the autoscaler to check whether
+    // a new node is fully initialized.
+    NodeReady(nodeName string) (bool, error)
+}
+```
+
+The underlying implementation can either be compiled into a custom autoscaler
+binary by cloud provider who controls the entire cluster or use HTTP similar to
+the [HTTP
+extender](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/extender.go#L41-L53).
+As an initial step, configuring such HTTP webhooks for different resource
+drivers can be added to the configuration file defined by the `--cloud-config`
+configuration file with a common field that gets added in all cloud provider
+configs or a new `--config` parameter can be added. Later, dynamically
+discovering deployed webhooks can be added through an autoscaler CRD.
```

Comment:
Minor heads up that we're beginning discussions about improving the config mechanism for the CA to stop the continuing explosion in command-line flags, current discussion focusing on use of configmaps.